ansible_base/lib/routers/associative_resource_router.py

```diff
@@ -0,0 +1,179 @@
+                {
+                    'instances': _('Cannot disassociate these objects because there must be a related object'),
+                }
+            )
+
+        return Response(status=status.HTTP_204_NO_CONTENT)
+
+    def get_serializer_class(self):
+        association_class = getattr(self, 'association_serializer', None)
+        if self.action in ('associate', 'disassociate'):
+            if association_class is None:
+                raise RuntimeError('You must set association_serializer on the viewset')
+            return association_class
+        return super().get_serializer_class()
+
+    def get_queryset(self):
+        parent_pk = self.kwargs['pk']
+        parent_model = self.backward_related_qs.model
+
+        try:
+            parent_instance = parent_model.objects.get(pk=parent_pk)
+        except parent_model.DoesNotExist:
+            return parent_model.objects.none()
+
+        child_queryset = getattr(parent_instance, self.association_fk).all()
+        return child_queryset
+
+
+class ReverseViewMixin:
+    def get_serializer_class(self):
+        related_class = getattr(self, 'related_serializer', None)
+        if self.action in ('associate', 'disassociate'):
+            if related_class is None:
+                raise RuntimeError('You must set related_serializer on the viewset')
+            return related_class
+        return super().get_serializer_class()
+
+    def get_queryset(self):
+        parent_pk = self.kwargs['pk']
+        parent_model = self.backward_related_qs.model
+
+        try:
+            parent_instance = parent_model.objects.get(pk=parent_pk)
+        except parent_model.DoesNotExist:
+            return parent_model.objects.none()
+
+        child_queryset = getattr(parent_instance, self.related_fk).all()
+        return child_queryset
+
+
+class AssociationResourceRouter(routers.SimpleRouter):
+    def get_method_map(self, viewset, method_map):
+        is_associate_viewset = issubclass(viewset, AssociateMixin)
+        associate_actions = ['associate', 'disassociate', 'list']
+        bound_methods = {}
+        for method, action_str in method_map.items():
+            if hasattr(viewset, action_str):
+                if is_associate_viewset and action_str not in associate_actions:
+                    continue
+                bound_methods[method] = action_str
+        return bound_methods
+
+    def register(self, prefix, viewset, basename=None, related_views={}, reverse_views={}):
+        if basename is None:
+            basename = self.get_default_basename(viewset)
+
+        for reverse_name, (reverse_view, fk) in reverse_views.items():
+
+            def related_serializer_factory(reverse_view=reverse_view):
+                class RelatedSerializer(serializers.Serializer):
+                    instances = serializers.PrimaryKeyRelatedField(
+                        queryset=reverse_view.queryset,
+                        many=True,
```

Comment:
from the docs

https://www.django-rest-framework.org/api-guide/relations/

> queryset - The queryset used for model instance lookups when validating the field input. Relationships must either set a queryset explicitly, or set read_only=True.

Shouldn't this be `read_only=True`, because the POST to associate is a totally separate and unrelated action?

Reply:
The screens render w/o the post options on them the way it is. From a quick read it sounds like adding `read_only: True` wouldn't hurt is probably better. However, when I added it on the POST option showed up. There may be some other black magic going on here. @relrod you have any ideas?

Reply:
Whether the POST option shows up is more the business of the viewset, not the serializer. I'd expect that the preferences of `reverse_view` dictates whether an object can be created, which isn't really right.

I'm getting confused by a number of things. The view `ReverseViewMixin` defines `get_queryset`, but... this doesn't fully add up. Is it the `instances` queryset that defines what list of related objects to show? Or is it the view `get_queryset`?

Reply:
Lets see what @relrod says, I mostly dug in deep enough to "make it work".

Reply:
Also, this function moved into another mixin to try and reduce untested duplicated code.